### PR TITLE
#5 Now sends to a slack channel instead of just logging it

### PR DESCRIPTION
**Pull request by Alex-Potter-Dixon**

### CODE REVIEW

File: .github/workflows/cd.yml [hunk 1]
```diff
@@ -9,28 +9,28 @@ jobs:
...
-        AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: "3.7"
+
+      - name: Install dependencies using pipenv
+        run: |
+          python -m pip install --upgrade pip pipenv
+          pipenv install --dev
```
Comment (Figglewatts) [inline .github/workflows/cd.yml:23]: Should be `pipenv sync --dev` for reproducable builds (uses the lockfile)

File: nsg_checker/message_dispatcher.py [hunk 2]
```diff
@@ -7,41 +7,87 @@
...
+    slack_oauth (str): The slack Oauth token.
+    slack_channel (str): The slack channel to send notifications to.
 
 """
 
 import logging
+from slack import WebClient
 
 
 class MessageDispatcher:
```
Comment (Figglewatts) [inline nsg_checker/message_dispatcher.py:19]: No class docstring?

File: nsg_checker/message_dispatcher.py [hunk 3]
```diff
@@ -7,41 +7,87 @@
...
     def __init__(self, o365_rules, gsuite_rules, o365_azure_rules,
```
Comment (Figglewatts) [inline nsg_checker/message_dispatcher.py:20]: No init typing or docstring? Needs types for variables and a `-> None` at the end

File: nsg_checker/message_dispatcher.py [hunk 4]
```diff
@@ -7,41 +7,87 @@
...
         self.missing_o365 = o365_rules.difference(o365_azure_rules)
         self.extra_o365 = o365_azure_rules - o365_rules
         self.missing_gsuite = gsuite_rules.difference(gsuite_azure_rules)
         self.extra_gsuite = gsuite_azure_rules - gsuite_rules
+        self.slack_client = WebClient(token=slack_oauth)
+        self.slack_channel = slack_channel
 
     def dispatch_message(self):
         """
         Dispatches the message to stdout. PLACEHOLDER for SNS work
```
Comment (Figglewatts) [inline nsg_checker/message_dispatcher.py:32]: Doesn't necessarily dispatch to stdout - it logs it which may or may not go to stdout depending on the config. Could be misleading for any user.

File: nsg_checker/message_dispatcher.py [hunk 5]
```diff
@@ -7,41 +7,87 @@
...
+                                           text=slack_text)
 
-        logging.info(
-            f"These NSG rules for GSuite are no longer needed: {self.extra_gsuite}"
-        )
+        logging.info(f"Sent slack message to {self.slack_channel}")
 
-        logging.info(
-            "Finished running Azure NSG Watcher Function. Closing down...")
+    def create_slack_message(self):
```
Comment (Figglewatts) [inline nsg_checker/message_dispatcher.py:46]: No return typing for function.

File: nsg_checker/message_dispatcher.py [hunk 6]
```diff
@@ -7,41 +7,87 @@
...
+
+        result = '\n'.join(
+            filter(None, [
+                missing_o365_message, missing_gsuite_message,
+                extra_o365_message, extra_gsuite_message
+            ]))
+
+        return result
+
+    def pretty_nsg_sets(self, nsg_set):
```
Comment (Figglewatts) [inline nsg_checker/message_dispatcher.py:80]: No return typing for function, or typing for argument.

File: handler.py [hunk 7]
```diff
@@ -33,7 +33,7 @@ def run(event, context):
     o365_rules = nsg_checker.get_o365_smtp_ipv4_cidrs()
     gsuite_rules = nsg_checker.get_gsuite_smtp_ipv4_cidrs()
     
-    dispatcher = MessageDispatcher(o365_rules,gsuite_rules,o365_azure_result, gsuite_azure_result)
+    dispatcher = MessageDispatcher(o365_rules,gsuite_rules,o365_azure_result, gsuite_azure_result, azure_credentials["slack_oauth"], os.environ["SLACK_CHANNEL"])
```
Comment (tpilvelis-gw) [inline handler.py:36]: I think it needs the channel id, not the channel name. would I be correct in saying it takes this env variable from Serverless

File: nsg_checker/message_dispatcher.py (hunk 2 above)
Comment (Alex-Potter-Dixon) [reply nsg_checker/message_dispatcher.py:19]: There is above it?

Comment (Figglewatts) [reply nsg_checker/message_dispatcher.py:19]: That's a module docstring, it won't appear in the autocomplete when using `MessageDispatcher`, only when importing `message_dispatcher`

File: handler.py (hunk 7 above)
Comment (Alex-Potter-Dixon) [reply handler.py:36]: Yeah, I assumed name, I'll change the docs to be ID. Cheers

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -8,45 +8,43 @@ on:
...
+      - uses: actions/checkout@v2
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v1
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install dependencies using pipenv
+        run: |
+          python -m pip install --upgrade pip pipenv
+          pipenv install --dev
```
Comment (Figglewatts) [inline .github/workflows/ci.yml:26]: Needs to be `pipenv sync --dev` for reproducability!

**Pull request merged**